Mathlib/Control/Monad/Free.lean

```diff
@@ -0,0 +1,481 @@
+
+This construction provides a free monad for any type constructor `f`, allowing for composable
+effect descriptions that can be interpreted later. Unlike the traditional Free monad,
+this does not require `f` to be a functor. -/
+inductive FreeM.{u, v, w} (f : Type u → Type v) (α : Type w) where
+  | protected pure : α → FreeM f α
+  | liftBind {ι : Type u} (op : f ι) (cont : ι → FreeM f α) : FreeM f α
+
+universe u v w
+
+namespace FreeM
+
+/-- Map a function over a `FreeM` monad. -/
+def map {α β : Type _} (F : Type u → Type v) (f : α → β) : FreeM F α → FreeM F β :=
+fun FFa =>
+  match FFa with
+  | .pure a => .pure (f a)
+  | .liftBind op cont => .liftBind op (fun z => FreeM.map F f (cont z))
+
+instance {F : Type u → Type v} : Functor (FreeM F) where
+  map := FreeM.map F
+
+instance {F : Type u → Type v} : LawfulFunctor (FreeM F) where
+  map_const {α β} := by
+    simp [Functor.mapConst, Functor.map]
+  id_map x := by
+    simp [Functor.map]
+    induction' x with a b op cont ih
+    · simp [map]
+    · simp [map, ih]
+  comp_map g h x := by
+    simp [Functor.map]
+    induction x
+    case pure a => simp [FreeM.map]
+    case liftBind op cont ih => simp [FreeM.map, ih]
+
+/-- Bind operation for the `FreeM` monad. -/
+protected def bind {a b : Type _} (F : Type u → Type v) (x : FreeM F a) (f : a → FreeM F b) :
+FreeM F b :=
+  match x with
+  | .pure a => f a
+  | .liftBind op cont => .liftBind op (fun z => FreeM.bind F (cont z) f)
+
+/-- Lift an operation from the effect signature `f` into the `FreeM f` monad. -/
+def lift {F : Type u → Type v} {ι : Type u} (op : F ι) : FreeM F ι :=
+  FreeM.liftBind op FreeM.pure
+
+@[simp]
+lemma lift_def {F : Type u → Type v} {ι : Type u} (op : F ι) :
+    lift op = FreeM.liftBind op FreeM.pure := by simp [lift]
+
+instance {F : Type u → Type v} : Monad (FreeM F) where
+  pure := FreeM.pure
+  bind := FreeM.bind F
+
+/-- The `liftBind` constructor is semantically equivalent to `(lift op) >>= cont`. -/
+lemma liftBind_eq_lift_bind {F : Type u → Type v} {ι : Type u} {α : Type u}
+    (op : F ι) (cont : ι → FreeM F α) :
+    FreeM.liftBind op cont = (lift op) >>= cont := by
+    simp [lift, bind, FreeM.bind]
+
+instance {F : Type u → Type v} : LawfulMonad (FreeM F) := by
+  apply LawfulMonad.mk'
+  · intro x y; induction' y with a b op cont ih'; all_goals {simp}
+  · intro _ _ x f; simp only [bind, FreeM.bind]
+  · intro _ _ _ x f g; induction' x with a b op cont ih;
+    · simp only [bind, FreeM.bind]
+    · simp only [bind, FreeM.bind, ih] at *; simp [ih]
+  · intro _ _ x y; simp [Functor.map, Functor.mapConst]
+  · intro _ _ x y; simp [bind, SeqLeft.seqLeft, Pure.pure]
+  · intro _ _ x y; simp [bind, SeqRight.seqRight]
+  · intro _ _ f x; induction' x with a b op cont ih;
+    · simp [bind, FreeM.bind, Functor.map, Pure.pure, map]
+    · simp only [Functor.map, bind, FreeM.bind, map, Pure.pure, map] at *; simp [ih]
+  · intro _ _ f x; simp [bind, FreeM.bind, Functor.map, Seq.seq]
```

Comment:
by using `apply` here you're defeating the point of `mk'`; look at some other examples of how to use it that avoid many of the subgoals.